File: socketMessenger.c

#include <stdio.h>
#include <sys/types.h>

#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <arpa/inet.h>
#include <time.h>

#include <signal.h>


// Instead of passing arguments, these objects can be static to be shared by following methods
int socketEndpoint;
struct sockaddr_in socketAddress;

// openConnection can be a client connection by Server, or the socketEndpoint by Client
int openConnection;


// For simpler structure, different running statuses are separated into methods
int runAsServer(void);
int runAsClient(char *hostingAddress);

// Since loops for writing and reading are same for Server & Client, they are encapsulated as methods
void writeToConnection(void);
void readFromConnection(void);


int main(int argc, const char * argv[]) {
	printf("\nSOCKET MESSENGER\n");

	// Initialize the socket. If initialization fails, end the process
	// Socket is configured to use IPv4 Internet protocols, Two-way TCP stream
	socketEndpoint = socket(AF_INET, SOCK_STREAM, 0);

	if (socketEndpoint < 0) {
		printf("\n Error : Could not create socket \n");
		return -1;
	}


	// Initialize the address struct.
	// Using hton(), change byte-order for any hosting address and port number 5000
	bzero(&socketAddress, sizeof(socketAddress));
	socketAddress.sin_family = AF_INET;
	socketAddress.sin_addr.s_addr = htonl(INADDR_ANY);
	socketAddress.sin_port = htons(5000);


	// This one program can run as Server or Client
	// At launch, the process evaluates arguments to see if it should run as server.
	if (argc > 1 && strcmp(argv[1], "server") == 0) {
		openConnection = runAsServer();
	}
	else {
		// The program assume that usual localhostIP is 127.0.0.1, or the user can change it by using arv[1]
		char *hostingAddress = (argc > 1) ? (char*)argv[1]:"127.0.0.1";
		openConnection = runAsClient(hostingAddress);
	}

	if (openConnection < 0) {
		return -1;
	}


	// getline() is needed for both Server and Client, but since it's a blocking method,
	// extra background process should run to read openConnection.
	// By using a 2 processes, reading and writing (by user) can operate concurrently from Server and Client

	pid_t pid = fork();

	// Reading is executed within the background process
	if (pid == 0) {
		readFromConnection();
	}
	else {
		writeToConnection();
	}


	// At this point, while-loop is broken, ready to terminate the process
	close(openConnection);

	// Kill the background process
	if (pid == 0) {
		printf("kill: %d\n", pid);
		kill(pid, SIGTERM);
	}

	return 0;
}


int runAsServer() {
	printf("RUN AS SERVER...\n");

	// Open socketEndpoint with assigned address and port number.
	// Wait for a client to connect by listening to socketEndpoint
	if (bind(socketEndpoint, (struct sockaddr*)&socketAddress, sizeof(socketAddress)) < 0) {
		printf("\n Error : Bind Failed \n");
		return -1;
	}


	// Listen to connection request with waiting queue of 1
	listen(socketEndpoint, 1);

	printf("\nLISTENING... \n");


	// When a client is connected, an endpoint is returned.
	int clientConnection = accept(socketEndpoint, (struct sockaddr*)NULL, NULL);

	if (clientConnection > 0) {
		time_t timestamp = time(NULL);
		printf("CLIENT %d CONNECTED AT: %.24s\n", clientConnection, ctime(&timestamp));
	}

	// Returned clientConnection(== openConnection) will be used for reading and writing.

	return clientConnection;
}

int runAsClient(char *hostingAddress) {
	printf("RUN AS CLIENT...\n");

	// Convert text address to integer binary form
	if (inet_pton(AF_INET, hostingAddress, &socketAddress.sin_addr) <= 0) {
		printf("\nhostingAddress: %s\n", hostingAddress);
		printf("\n inet_pton error occured\n");

		return -1;
	}


	// Client attempts connecting to hostingAddress. If connecting fails, end the process
	if (connect(socketEndpoint, (struct sockaddr *)&socketAddress, sizeof(socketAddress)) < 0) {
		printf("\n Error : Connect Failed \n");
		printf("\nTo run as server, type: \"./socketMessenger.out server\"\n");

		return -1;
	}


	time_t timestamp = time(NULL);
	printf("CONNECTED TO SERVER %s AT: %.24s\n", hostingAddress, ctime(&timestamp));


	// Returned socketEndpoint(== openConnection) will be used for reading and writing.

	return socketEndpoint;
}


void writeToConnection(void) {

	printf("\nSTART ENTERING MESSAGES\n(Type \"CLOSE\" to end connection):\n\n");

	// Repeat waiting for user to enter a line, sending it to connection
	while(1) {
		fputs("THIS USER: ", stdout);	// A prompt
		
		char *message = NULL;
		size_t bufferSize = 0;
		ssize_t inputSize = getline(&message, &bufferSize, stdin);

		// Skip, in case of just hitting return \n
		if (inputSize <  1 || strcmp(message, "\n") == 0) {
			continue;
		}


		// Write strlen of characters to openConnection
		write(openConnection, message, strlen(message));


		// Use "CLOSE" as a command, breaking the loop and return back to main()
		if (strstr(message, "CLOSE")) {
			break;
		}
	}
}

void readFromConnection(void) {

	while (1) {
		// Try reading set length from connection, though it may include trailing zeros
		char message[1024];
		bzero(message, sizeof(message));

		ssize_t messageSize = read(openConnection, message, sizeof(message)-1);

		if (messageSize < 0) {
			fputs("\n Read error \n", stdout);
			break;
		}


		// Disconnect trailing zeros by nullifying the end of the string
		message[messageSize] = 0;


		// Use "CLOSE" as a command, breaking the loop and return back to main()
		// If trimmed string length is actually 0, also break the loop
		if (strstr(message, "CLOSE") || strlen(message) <= 0) {
			break;
		}


		// To distinguish the received message from local messages,
		// a identification prompt is added in front of the string.
		char formatted[1024];
		bzero(formatted, sizeof(formatted));
		snprintf(formatted, sizeof(formatted), "\nFROM connection %d: %s", openConnection, message);

		// Print formatted message to standard output
		if (fputs(formatted, stdout) == EOF) {
			printf("\n Error : Fputs error\n");
		}

		// Because receiving end was also running as writer,
		// the prompt should be restored after printing received message.
		fputs("THIS USER: ", stdout);
		fflush(stdout);
	}
}

File: socketMessenger.py
import os
import sys
import socket

import argparse
import time, datetime
import logging
from multiprocessing import Process, Queue, Lock


global OPEN_SOCKET
OPEN_SOCKET = socket.socket()
OPEN_HOST = socket.gethostname()
OPEN_PORT = 12345

global CLIENT_CONNECTION, CLIENT_ADDRESS
CLIENT_CONNECTION = None
CLIENT_ADDRESS = None

global ACTIVE_CONNECTION
ACTIVE_CONNECTION = None

MESSAGE_BUFFER = 4096


def passiveReceivingWorker():

    while True:
        try:
            received = ACTIVE_CONNECTION.recv(MESSAGE_BUFFER)

            if len(received) > 0:
                print "\nReceived (", len(received), ") : ", received
                print "Message: ", ""
            else:
                break

        except KeyboardInterrupt, e:
            print e
            break

    print "ACTIVE CONNECTION: closed (Shutdown with Ctrl+C)"
    sys.exit(1)


if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG)
    print "\n[[SOCKET MESSENGER]]\n"

    ap = argparse.ArgumentParser()
    ap.add_argument("-s", "--server", required=True, help="Decide to run as server")

    args = vars(ap.parse_args())

    #print "args: ", args, "\n"


    SHOULD_RUN_AS_SERVER = ((args['server'] == "True") | (args['server'] == "true") | (args['server'] == "TRUE"))
    #print "SHOULD_RUN_AS_SERVER: ", SHOULD_RUN_AS_SERVER

    if SHOULD_RUN_AS_SERVER:
        print "[[ RUN AS SERVER... ]]\n"
        OPEN_SOCKET.bind((OPEN_HOST, OPEN_PORT))
        OPEN_SOCKET.listen(5)  # Now wait for client connection.

        while True:
            try:
                print "Waiting for a client..."

                CLIENT_CONNECTION, CLIENT_ADDRESS = OPEN_SOCKET.accept()  # Establish connection with client.
                ACTIVE_CONNECTION = CLIENT_CONNECTION

                print "Got connection from:", CLIENT_ADDRESS
                ACTIVE_CONNECTION.send('Thank you for connecting')
                break

            except KeyboardInterrupt, e:
                print e
                break

    else:
        print "[[ RUN AS CLIENT... ]]\n"

        OPEN_SOCKET.connect((OPEN_HOST, OPEN_PORT))
        ACTIVE_CONNECTION = OPEN_SOCKET


    PASSIVE_RECEIVING_PROCESS = Process(target=passiveReceivingWorker, args=())
    PASSIVE_RECEIVING_PROCESS.daemon = True
    PASSIVE_RECEIVING_PROCESS.start()


    while True:
        try:
            message = raw_input("Message:\n")
            ACTIVE_CONNECTION.send(message)

        except KeyboardInterrupt, e:
            print e
            break


    print "\n\n[[ FINISH RUNNING... ]]\n"

    PASSIVE_RECEIVING_PROCESS.join()
    PASSIVE_RECEIVING_PROCESS.terminate()
    #print "DID JOIN?"


    if ACTIVE_CONNECTION is not None:
        ACTIVE_CONNECTION.close()

    if CLIENT_CONNECTION is not None:
        CLIENT_CONNECTION.close()  # Close the client connection from server

    OPEN_SOCKET.close()  # Close the socket when done
